src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionSecurityTokenAuthenticator.cs

```diff
@@ -0,0 +1,1366 @@
+                isSenderFault = false;
+            }
+            else if (e is EndpointNotFoundException)
+            {
+                // send a receiver fault so that the sender can retry
+                subCode = new FaultCode(AddressingStrings.EndpointUnavailable, request.Version.Addressing.Namespace);
+                reason = new FaultReason(SR.Format(SR.SecurityListenerClosingFaultReason), CultureInfo.CurrentCulture);
+                isSenderFault = false;
+            }
+            else
+            {
+                subCode = new FaultCode(TrustApr2004Strings.InvalidRequestFaultCode, TrustFeb2005Strings.Namespace);
+                reason = new FaultReason(SR.Format(SR.InvalidRequestTrustFaultCode), CultureInfo.CurrentCulture);
+                isSenderFault = true;
+            }
+            FaultCode faultCode;
+            if (isSenderFault)
+            {
+                faultCode = FaultCode.CreateSenderFaultCode(subCode);
+            }
+            else
+            {
+                faultCode = FaultCode.CreateReceiverFaultCode(subCode);
+            }
+            MessageFault fault = MessageFault.CreateFault(faultCode, reason);
+            Message faultReply = Message.CreateMessage(request.Version, fault, request.Version.Addressing.DefaultFaultAction);
+            faultReply.Headers.RelatesTo = request.Headers.MessageId;
+            return faultReply;
+        }
+
+        void NotifyOperationCompletion(SecuritySessionOperation operation, SecurityContextSecurityToken newSessionToken, SecurityContextSecurityToken previousSessionToken, EndpointAddress remoteAddress)
+        {
+            if (operation == SecuritySessionOperation.Issue)
+            {
+                if (this.issuedSecurityTokenHandler != null)
+                {
+                    this.issuedSecurityTokenHandler(newSessionToken, remoteAddress);
+                }
+                else
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new SecurityNegotiationException(SR.IssueSessionTokenHandlerNotSet));
+                }
+            }
+            else if (operation == SecuritySessionOperation.Renew)
+            {
+                if (this.renewedSecurityTokenHandler != null)
+                {
+                    this.renewedSecurityTokenHandler(newSessionToken, previousSessionToken);
+                }
+                else
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new SecurityNegotiationException(SR.RenewSessionTokenHandlerNotSet));
+                }
+            }
+            else
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException());
+            }
+        }
+
+        public override Task CloseAsync(CancellationToken token)
+        {
+           /* TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
```

Comment:
We should be closing the SecureConversation service dispatcher here. rst means Request Security Token, so rstListener is the component which handles the creation of a new SecureConversation session.

Reply:
we are doing that in SecuritySessionServerSettings (in old way same https://referencesource.microsoft.com/#System.ServiceModel/System/ServiceModel/Security/SecuritySessionServerSettings.cs,8ab448e441b8781f ). The RstListner used to be an object (SecuritySessionHost) inherited from ServiceHost, which we discussed and concluded to keep as simple object without inheriting from ServiceHost... so we don't have to deal with open/close.. if you are worried about the object not being purged from memory,  I can add IDisposable interface and call explicitly from CloseAsync.. let me know what you think... keeping this open